torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -440,10 +440,16 @@ void DataChannelNccl::allReduce(std::vector<at::Tensor>& data,
                              ncclOp[operation],
                              (*comms)[i],
                              stream));
-    THCudaCheck(cudaEventRecord((*events)[i], stream));
   }
   NCCL_CHECK(ncclGroupEnd());
 
+  for (size_t i = 0; i < data.size(); ++i) {
+    auto device = data[i].get_device();
+    gpuGuard.setDevice(device);
```

Comment:
Does `cudaEventRecord` really need you to be on this particular device? Maybe we could just loop over input tensors, get their devices and respective streams, and then record the events?

Reply:
I assume it doesn't need to be, I just did this for safety purpose I guess